File: nim/tarai.nim

proc tarai(x, y, z: int): int =
  if x <= y:
    return y
  else:
    return tarai(tarai(x - 1, y, z), tarai(y - 1, z, x), tarai(z - 1, x, y))

echo tarai(12, 6, 0)

File: pascal/tarai.pas
program tarai;

function tarai(x, y, z: integer): integer;
begin
  if x <= y then
    tarai := y
  else
    tarai := tarai(tarai(x - 1, y, z), tarai(y - 1, z, x), tarai(z - 1, x, y));
end;

begin
  writeln(tarai(12, 6, 0));
end.
